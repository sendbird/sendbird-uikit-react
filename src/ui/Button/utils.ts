import { ButtonTypes, ButtonSizes } from './types';

export function changeTypeToClassName(type: ButtonTypes): string {
  switch (type) {
    case ButtonTypes.PRIMARY: return 'sendbird-button--primary';
    case ButtonTypes.SECONDARY: return 'sendbird-button--secondary';
    case ButtonTypes.DANGER: return 'sendbird-button--danger';
    case ButtonTypes.DISABLED: return 'sendbird-button--disabled';
    case ButtonTypes.WARNING: return 'sendbird-button--warning';

    default: return null;
  }
}

export function changeSizeToClassName(size: ButtonSizes): string {
  switch (size) {
    case ButtonSizes.BIG: return 'sendbird-button--big';
    case ButtonSizes.SMALL: return 'sendbird-button--small';
    default: return null;
  }
}
