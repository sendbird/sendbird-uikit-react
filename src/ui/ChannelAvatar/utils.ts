import type { GroupChannel } from "@sendbird/chat/groupChannel";
import type { OpenChannel } from "@sendbird/chat/openChannel";

export const DEFAULT_URL_PREFIX = 'https://static.sendbird.com/sample/cover/cover_';

export const getOpenChannelAvatar = (channel: OpenChannel): string => {
  if (channel?.coverUrl) {
    return channel.coverUrl;
  }
};

export const getChannelAvatarSource = (channel: GroupChannel, currentUserId: string): string | Array<string> => {
  if (channel?.coverUrl) {
    if (!(new RegExp(`^${DEFAULT_URL_PREFIX}`).test(channel.coverUrl))) {
      return channel.coverUrl;
    }
  }
  return (channel?.members || [])
    .filter((member) => member.userId !== currentUserId)
    .map(({ profileUrl }) => profileUrl);
};

export const generateDefaultAvatar = (channel: GroupChannel): boolean => {
  if (channel?.coverUrl) {
    if (new RegExp(`^${DEFAULT_URL_PREFIX}`).test(channel.coverUrl)) {
      return true;
    }
    return false;
  }
  return true;
}
