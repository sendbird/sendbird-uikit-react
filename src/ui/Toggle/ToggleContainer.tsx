import React, { ChangeEvent, useCallback, useState } from 'react';

import {
  ToggleContext,
  ToggleContextInterface,
  TOGGLE_DEFAULT_VALUE as Dvalue,
} from './ToggleContext';

export interface ToggleContainerProps extends ToggleContextInterface {
  children?: React.ReactElement | Array<React.ReactElement>;
}

// Props Explanation https://github.com/aaronshaf/react-toggle#props
export function ToggleContainer({
  checked = Dvalue.checked, // null
  defaultChecked = Dvalue.defaultChecked,
  disabled = Dvalue.disabled,
  onChange = Dvalue.onChange,
  onFocus = Dvalue.onFocus,
  onBlur = Dvalue.onBlur,
  children,
}: ToggleContainerProps): React.ReactElement {
  const [isChecked, setChecked] = useState(defaultChecked || false);
  const handleChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    if (disabled) {
      return;
    }
    if (checked === null) {
      setChecked(e.currentTarget.checked);
    }
    onChange(e);
  }, [onChange, checked]);

  return (
    <ToggleContext.Provider value={{
      checked: checked !== null ? checked : isChecked,
      disabled,
      onChange: handleChange,
      onFocus: (e) => {
        if (!disabled) {
          onFocus(e);
        }
      },
      onBlur: (e) => {
        if (!disabled) {
          onBlur(e);
        }
      },
    }}>
      {children}
    </ToggleContext.Provider>
  );
}
