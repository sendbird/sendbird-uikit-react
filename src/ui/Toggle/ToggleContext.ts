import React, { ChangeEventHandler } from 'react';

/**
 * The default value of `checked` should be null
 * to support both case of controlled and uncontrolled component
 * ref: https://github.com/aaronshaf/react-toggle#props
 */
const noop = ():void => { /* noop */ };
export const TOGGLE_DEFAULT_VALUE = {
  checked: null,
  defaultChecked: false,
  disabled: false,
  reversed: false,
  onChange: noop,
  onFocus: noop,
  onBlur: noop,
};

export interface ToggleContextInterface {
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  reversed?: boolean;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onFocus?: ChangeEventHandler<HTMLInputElement>;
  onBlur?: ChangeEventHandler<HTMLInputElement>;
}

export const ToggleContext = React.createContext<ToggleContextInterface>(TOGGLE_DEFAULT_VALUE);

export function useToggleContext(): ToggleContextInterface {
  const context = React.useContext(ToggleContext);
  return context;
}
