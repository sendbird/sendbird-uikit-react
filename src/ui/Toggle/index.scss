@import '../../styles/variables';

.sendbird-input-toggle-button {
  position: relative;
  display: inline-flex;
  align-items: center;

  box-sizing: border-box;
  cursor: pointer;
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--checked {
  @include themed() {
    background-color: t(primary-3);
    border: 1px solid t(primary-3);
  }
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--unchecked {
  @include themed() {
    background-color: t(bg-3);
    border: 1px solid t(bg-3);
  }
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--disabled {
  cursor: not-allowed;
  @include themed() {
    background-color: t(bg-2);
    border: 1px solid t(bg-2);
  }
}

.sendbird-input-toggle-button input {
  position: absolute;
  display: none;
}

.sendbird-input-toggle-button__inner-dot {
  position: absolute;
  border-radius: 50%;
  @include themed() {
    background-color: t(bg-0);
  }
  animation-iteration-count: 1;
}

/* Manage animation and position by status */
@keyframes moveToRight {
  0% { right: 60%; }
  100% { right: 10%; }
}
@keyframes moveToLeft {
  0% { right: 10%; }
  100% { right: 60%; }
}
// normal - animation
.sendbird-input-toggle-button.sendbird-input-toggle-button--turned-on .sendbird-input-toggle-button__inner-dot {
  animation-name: moveToRight;
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--turned-off .sendbird-input-toggle-button__inner-dot {
  animation-name: moveToLeft;
}
// reverse - animation
.sendbird-input-toggle-button.sendbird-input-toggle-button--reversed.sendbird-input-toggle-button--turned-on .sendbird-input-toggle-button__inner-dot {
  animation-name: moveToLeft;
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--reversed.sendbird-input-toggle-button--turned-off .sendbird-input-toggle-button__inner-dot {
  animation-name: moveToRight;
}
// normal - position
.sendbird-input-toggle-button.sendbird-input-toggle-button--unchecked .sendbird-input-toggle-button__inner-dot {
  right: 60%;
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--checked .sendbird-input-toggle-button__inner-dot {
  right: 10%;
}
// reverse - position
.sendbird-input-toggle-button.sendbird-input-toggle-button--unchecked.sendbird-input-toggle-button--reversed .sendbird-input-toggle-button__inner-dot {
  right: 10%;
}
.sendbird-input-toggle-button.sendbird-input-toggle-button--checked.sendbird-input-toggle-button--reversed .sendbird-input-toggle-button__inner-dot {
  right: 60%;
}
