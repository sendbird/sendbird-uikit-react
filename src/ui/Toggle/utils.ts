import useSendbirdStateContext from '../../hooks/useSendbirdStateContext';

export function filterNumber(input: string | number): Array<number> {
  if (typeof input !== 'string' && typeof input !== 'number') {
    try {
      const { config } = useSendbirdStateContext();
      const { logger } = config;
      logger.warning('@sendbird/uikit-react/ui/Toggle: TypeError - expected string or number.', input);
    } catch (_) { /* noop */ }
    return [];
  }
  if (typeof input === 'number') {
    return [input];
  }
  const regex = /(-?\d+)(\.\d+)?/g;
  const numbers = input.match(regex) || [];
  return numbers.map(parseFloat);
}
