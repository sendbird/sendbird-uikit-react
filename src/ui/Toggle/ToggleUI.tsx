import React, { useState } from 'react';
import { useToggleContext } from './ToggleContext';
import { filterNumber } from './utils';

export interface ToggleUIProps {
  reversed?: boolean;
  width?: string;
  animationDuration?: string;
  style?: Record<string, string>;
  name?: string;
  id?: string;
  ariaLabel?: string;
  ariaLabelledby?: string;
}

export function ToggleUI(props: ToggleUIProps): React.ReactElement {
  const {
    reversed = false,
    width = '40px',
    animationDuration = '0.5s',
    style = {},
    name = '',
    id = '',
    ariaLabel = '',
    ariaLabelledby = '',
  } = props;
  const {
    checked,
    disabled,
    onChange,
    onFocus,
    onBlur,
  } = useToggleContext();

  // animation should not be activated in the initialization step
  const [animatedClassName, setAnimatedClassName] = useState('');

  const toggleWidth = filterNumber(width)?.[0];
  const toggleHeight = toggleWidth / 2;
  // The size of dot should be 60% of toggle height
  const dotSize = toggleHeight * 0.6;

  return (
    <label
      className={[
        'sendbird-input-toggle-button',
        animatedClassName,
        checked ? 'sendbird-input-toggle-button--checked' : 'sendbird-input-toggle-button--unchecked',
        disabled ? 'sendbird-input-toggle-button--disabled' : [],
        reversed ? 'sendbird-input-toggle-button--reversed' : [],
      ].flat().join(' ')}
      style={{
        width: `${toggleWidth}px`,
        height: `${toggleHeight}px`,
        borderRadius: `${dotSize}px`,
        ...style,
      }}
    >
      <div
        className={[
          'sendbird-input-toggle-button__inner-dot',
          checked
            ? 'sendbird-input-toggle-button__inner-dot--activate'
            : 'sendbird-input-toggle-button__inner-dot--inactivate',
        ].join(' ')}
        style={{
          width: `${dotSize}px`,
          height: `${dotSize}px`,
          animationDuration,
        }}
      />
      <input
        type="checkbox"
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledby}
        name={name}
        id={id}
        checked={checked ?? undefined}
        disabled={disabled}
        onChange={(e) => {
          onChange?.(e);
          setAnimatedClassName(e.currentTarget.checked ? 'sendbird-input-toggle-button--turned-on' : 'sendbird-input-toggle-button--turned-off');
        }}
        onFocus={onFocus}
        onBlur={onBlur}
      />
    </label>
  );
}
