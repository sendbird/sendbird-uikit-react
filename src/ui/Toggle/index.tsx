import React from 'react';
import './index.scss';

import {
  ToggleContainer as _ToggleContainer,
  ToggleContainerProps as _ToggleContainerProps,
} from './ToggleContainer';
import { useToggleContext as _useToggleContext } from './ToggleContext';
import {
  ToggleUI as _ToggleUI,
  ToggleUIProps as _ToggleUIProps,
} from './ToggleUI';

// export - context
export const ToggleContainer = _ToggleContainer;
export type ToggleContainerProps = _ToggleContainerProps;
export const useToggleContext = _useToggleContext;
// export - UI
export const ToggleUI = _ToggleUI;
export type ToggleUIProps = _ToggleUIProps;

// export - Toggle
export interface ToggleProps extends _ToggleContainerProps, _ToggleUIProps {
  className?: string;
}
export default function Toggle(props: ToggleProps): React.ReactElement {
  const {
    // ToggleProvider
    checked,
    defaultChecked,
    disabled,
    reversed,
    onChange,
    onFocus,
    onBlur,
    // ToggleUI
    className,
    width,
    // height will be half of width
    animationDuration,
    style,
    name,
    id,
    ariaLabel,
    ariaLabelledby,
  } = props;
  return (
    <div className={`sendbird-ui-toggle ${className}`}>
      <_ToggleContainer
        checked={checked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        reversed={reversed}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
      >
        <_ToggleUI
          width={width}
          animationDuration={animationDuration}
          style={style}
          name={name}
          id={id}
          ariaLabel={ariaLabel}
          ariaLabelledby={ariaLabelledby}
        />
      </_ToggleContainer>
    </div>
  );
}
