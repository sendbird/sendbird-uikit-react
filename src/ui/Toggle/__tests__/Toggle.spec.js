import React from 'react';
import { render } from '@testing-library/react';

import Toggle from "../index";
import { filterNumber } from '../utils';

describe('Toggle', () => {
  it('should do a snapshot test of the Toggle DOM', function() {
    const { asFragment } = render(<Toggle />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Toggle/utils: filterNumber', () => {
  it('should filter int from string', () => {
    expect(filterNumber('abcd567hijk')[0]).toBe(567);
  });
  it('should filter float from string', () => {
    expect(filterNumber('abcd5.67hijk')[0]).toBe(5.67);
  });
  it('should filter several int', () => {
    expect(filterNumber('abcd5 6 7hijk')).toStrictEqual([5, 6, 7]);
  });
  it('should filter several float', () => {
    expect(filterNumber('abcd5.6.7hijk')).toStrictEqual([5.6, 7]);
  });
  it('should filter negative numbers', () => {
    expect(filterNumber('abcd-1 2-3.4hijk')).toStrictEqual([-1, 2, -3.4]);
  });
});
