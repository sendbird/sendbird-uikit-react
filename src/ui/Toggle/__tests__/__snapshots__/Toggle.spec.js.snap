// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should do a snapshot test of the Toggle DOM 1`] = `
<DocumentFragment>
  <div
    class="sendbird-ui-toggle undefined"
  >
    <label
      class="sendbird-input-toggle-button  sendbird-input-toggle-button--unchecked"
      data-testid="sendbird-input-toggle-button"
      style="width: 40px; height: 20px; border-radius: 12px;"
    >
      <div
        class="sendbird-input-toggle-button__inner-dot sendbird-input-toggle-button__inner-dot--inactivate"
        style="width: 12px; height: 12px; animation-duration: 0.5s;"
      />
      <input
        aria-label=""
        aria-labelledby=""
        id=""
        name=""
        type="checkbox"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`Toggle should do a snapshot test of the Toggle DOM with props 1`] = `
<DocumentFragment>
  <div
    class="sendbird-ui-toggle mock-class-name"
  >
    <label
      class="sendbird-input-toggle-button  sendbird-input-toggle-button--checked sendbird-input-toggle-button--disabled sendbird-input-toggle-button--reversed"
      data-testid="sendbird-input-toggle-button"
      style="width: 80px; height: 40px; border-radius: 24px; border: 1px solid red;"
    >
      <div
        class="sendbird-input-toggle-button__inner-dot sendbird-input-toggle-button__inner-dot--activate"
        style="width: 24px; height: 24px; animation-duration: 0.1s;"
      />
      <input
        aria-label="mock-element-aria-label"
        aria-labelledby="mock-element-aria-labeled-by"
        checked=""
        disabled=""
        id="mock-element-id"
        name="mock-element-name"
        type="checkbox"
      />
    </label>
  </div>
</DocumentFragment>
`;
